### PR TITLE
Tree-sitter step 1: vendor runtime lib + add lua API

### PR DESCRIPTION
**Pull request by bfredl**
#9219 has already grown pretty large. To get something reviewable and mergable in a resonable timeframe, I will try to divide it into smaller PRs. My suggestion for a MVP is the following: vendor the tree-sitter [runtime lib](https://github.com/tree-sitter/tree-sitter/tree/master/lib/src)  into our source-tree, and expose syntax trees as a lua API. More specific features such as syntax highlighting and multi-language parsing will be follow up PRs.

Vendoring the runtime allows us to always build the lua tree-sitter bindings and support code (there will be no `FEAT_TREESITTER`), without making the build any more complicated for distributors. The runtime lib without parsers is not comparatively large, about as big as nvim's `syntax.c` in its entirety. Though I expect `libutf8proc` to be common in distros and "native binary" package solutions already, so we should be able to treat that as an ordinary dependency.

On the other hand, the tree-sitter parsers for specific languages will  initially _not_ be built by the nvim build system. Instead we expect those who experiment with the MVP to e.g. build the parsers they want themselves for their respective platform, and nvim will provide an interface to load them as dynamic libraries. This means we can discuss and work on a reasonable final distribution strategy _in parallel_ with other tasks (as mentioned above). 

CC @maxbrunsfeld if you have any comments of our vendoring strategy.

### CODE REVIEW

**Comment by maxbrunsfeld** [comment]
> Vendoring the runtime

> nvim will provide an interface to load them as dynamic libraries.

👍 This makes sense to me.

In case it's useful as a reference, the [Tree-sitter CLI](https://github.com/tree-sitter/tree-sitter/tree/master/cli) has [logic](https://github.com/tree-sitter/tree-sitter/blob/dbecdf6c2e77a3e3cf756c710e2627f30b68bbf2/cli/src/loader.rs#L159-L190) for automatically compiling Tree-sitter parsers into dynamic libraries and dynamically loading them when you run the `tree-sitter test` and `tree-sitter parse` commands.

Great to see this happening! Let me know if I can help with anything.

**Comment by bfredl** [comment]
> In case it's useful as a reference, the Tree-sitter CLI has logic for automatically compiling Tree-sitter parsers into dynamic libraries and dynamically loading them when you run the tree-sitter test and tree-sitter parse commands.

That sounds nice. Maybe we can use it to automate building of parsers for our own binary packages (like nvim.appimage) later on.

**Comment by phodge** [comment]
Does this main tree-sitter will be doing all its work in the main thread? And if so, would it be hard to push this processing work out into a child process using buffer update events and RPC in the future?

**Comment by bfredl** [comment]
@phodge the idea is is that most parses would be really quick. Thus we benefit from the simplicity of accessing the tree-sitter node API in-process with a thin lua wrapper. I think atom uses a hybrid method, it parses in the main thread up to an operation limit, but then pushes the remaining parsing job to a separate thread if it is too slow. It would be interesting to see if we can do something similar.

**Comment by phodge** [comment]
> the idea is is that most parses would be really quick. Thus we benefit from the simplicity of accessing the tree-sitter node API in-process with a thin lua wrapper.

I understand that Tree-sitter is designed to be fast enough for running in the main thread, and that it makes the APIs much simpler if we do it in the main thread. But I think we should be cautious about this approach because these are both reasons why Vim failed us and we had to fork Neovim. Too many things competing for a single CPU core, and no way to distribute the workload because IPC is hard. But tree-sitter in a separate process is probably a problem for the future.

If it's not too much to ask, I'd love to see a help text explaining how to use the Lua functions to access the syntax tree. It looks like currently all I can do is turn on tree-sitter for a buffer and then ask TS to highlight the node under the cursor?

**Comment by bfredl** [comment]
> If it's not too much to ask, I'd love to see a help text explaining how to use the Lua functions to access the syntax tree. It looks like currently all I can do is turn on tree-sitter for a buffer and then ask TS to highlight the node under the cursor?

Sure, apart form fixing the build system, the plan is to add docs and tests for part of the API we will expose in this first PR :)

**Comment by maxbrunsfeld** [comment]
> it parses in the main thread up to an operation limit, but then pushes the remaining parsing job to a separate thread if it is too slow. It would be interesting to see if we can do something similar.

Yeah, Atom uses [this API](https://github.com/tree-sitter/tree-sitter/blob/dbecdf6c2e77a3e3cf756c710e2627f30b68bbf2/lib/include/tree_sitter/api.h#L238) to set a timeout of ~1 millisecond for main thread parsing. And if that timeout expires, it resumes parsing on a background thread. The syntax trees are persistent, so they can be cheaply [copied](https://github.com/tree-sitter/tree-sitter/blob/dbecdf6c2e77a3e3cf756c710e2627f30b68bbf2/lib/include/tree_sitter/api.h#L298) for use an a background thread.

I'm curious how this would translate to Neovim. In Atom's case, it relies on the text buffer having a *snapshot* API that allows you to safely read from a certain version of the buffer in a background thread, even if the user is mutating the buffer on the main thread. Does the vim buffer have a way to do something like that?

**Comment by bfredl** [comment]
> I'm curious how this would translate to Neovim. In Atom's case, it relies on the text buffer having a snapshot API that allows you to safely read from a certain version of the buffer in a background thread, even if the user is mutating the buffer on the main thread. Does the vim buffer have a way to do something like that?

There is some form of "locking" mechanism to lock parts of the buffer  in the backing storage (memline.c), but I don't think it was ever intended for safe multithreaded usage :P We probably have to make a straight copy of the modified part of the buffer in that case. But `ts_tree_copy` would still be a reason to consider multi-threading and not only multi-processing.

**Comment by phodge** [comment]
> I'm curious how this would translate to Neovim.

Neovim has an API for sending buffer change deltas to a child process, so inside a child process you could let the change deltas queue up while TS finishes parsing the buffer in some known state. But I don't think that helps us in terms of doing the first part of the processing in the main Neovim process because you'd have to serialize all the completed syntax nodes and send them out to the child process via RPC before the child process can finish the work.

AFAIK neovim is completely single-threaded with a focus on using RPC and child processes to achieve concurrency.

**Comment by justinmk** [comment]
> Too many things competing for a single CPU core, and no way to distribute the workload because IPC is hard

https://github.com/neovim/neovim/pull/9943 is exploring ways to make IPC easier.

**Comment by justinmk** [comment]
bikeshed: would be nice if "tree sitter" is spelled `treesitter` everywhere (filenames etc.), so that searching for it is more predictable.

**Comment by bfredl** [comment]
Update: _basic_ tree-sitter test now works on travis (64-bit linux and OS X). windows/appveyor still requires some work...

**Comment by bfredl** [comment]
Building the parser and running tests now work on 32-bit Appveyor (MINGW_64 nvim also compiles, but the parser `.dll` built by tree-sitter-cli is 32-bit and won't load).

**Comment by bfredl** [comment]
@justinmk would it be possible to install 64-bit node instead on the 64-bit appveyor builds? Or should we just skip tree-sitter tests on MINGW_64 for now? (We can look into compiling the languages by ourselves later, node is not really required).

**Comment by maxbrunsfeld** [comment]
> the parser .dll built by tree-sitter-cli is 32-bit and won't load

> would it be possible to install 64-bit node instead on the 64-bit appveyor builds?

If for some reason you don't want to change the Node installation, you could also download the 64-bit `tree-sitter` binary directly [from GitHub](https://github.com/tree-sitter/tree-sitter/releases/tag/0.15.5), instead of installing it with `npm`. When you install it with `npm`, it just detects the platform and architecture and downloads the appropriate binary from the GitHub release.

**Comment by bfredl** [comment]
@maxbrunsfeld Thanks. that would cover all travis/appveyor builds except 32-bit linux. I guess we could bulid that binary ourselves and keep it in the cached dependencies, for now.

**Comment by bfredl** [comment]
Tests now work on every platform (except 32-bit linux, currently hacking on that).

I will now start writing some proper documentation, but here is a quick overview how it currently can be used.

1. Currently the tester/user will have to build the parser themselves, for instance using the tree-sitter cli utility. The parser is loaded into nvim using ("c" as an example, but any language should work)

```lua
    vim.treesitter.add_language("/path/to/c_parser.so", "c")
```

This will be more convenient later (like a standard path for finding parsers, our own binary distribution, etc).

2. Create a parser for a buffer and a given language (if another plugin uses the same buffer/lang combination, it will be safely reused). Use

```lua
    parser = vim.treesitter.get_parser(bufnr, lang)
```

`bufnr=0` can be used for current buffer. `lang` will default to `&filetype` (doesn't work for every language, we will need to map the names). Currently the parser will be retained for the lifetime of a buffer but this might change.  A plugin should keep a reference to the parser as long as it is interested in incremental updates.

3. Whenever you need to access the current syntax tree, parse the buffer.

```lua
    tree = parser:parse()
```

This will return an _immutable_ tree that represents the current state of the buffer. Whenever the plugin wants to access the state after a (possible) edit it should call `parse()` . This doesn't do more work than necessary, if the buffer wasn't edited, the same tree will be returned again. if the buffer was parsed before, incremental parsing will be done of the changed parts (TODO: #10190 will be needed to make this efficient for small edits, but this will not change the interface)

(NB: if you use the parser directly inside a `nvim_buf_attach` callback, you must call `get_parser` _before_ you register your callback. But probably you want to use a timer or something anyway, to throttle reparsing and whatever work the plugin does on the tree).

4. Explore/use the tree. We (plan to) expose most of the (non-mutating) methods of the C API, as documented at https://tree-sitter.github.io/tree-sitter/using-parsers#retrieving-nodes and https://github.com/tree-sitter/tree-sitter/blob/83f0ea80cf6d474ab92f1b68fde0142f0d782faa/lib/include/tree_sitter/api.h#L350

```lua
    root = tree:root() -- get the root node
    node = root:descendant_for_point_range(row,col,row,col) -- leaf node touching the cursor (row,col)
    parent = node:parent() -- parent of this node
    print(parent:type(), parent:child_count()) -- type and child count
    first = parent:child(0) - NB: zero indexed like the C API
    row1, col1, row2, col2 = parent:range() -- bounding range, zero-indexed rows/cols
```

**Comment by bfredl** [comment]
As a heads-up, we have decided not to merge this for 0.4. Our priority for 0.4 will be to fix loose ends in already merged features, which already are plenty (multigrid/floats, lua event loop and stdlib, external UI features, API/eval additions, etc ). It is better to postpone treesitter to 0.5 where we can spend the entire cycle in implementing high-level treesitter features (such as highlighting, indentation, etc).

That being said, I will keep this branch rebased (which is mostly frictionless, as this code is quite isolated), and it should be fully functional to anyone who wants to experiment with treesitter . The "step 1" API support should be good enough for features such as "smart" movement and selection. I might still decide to do some small API adjustments (like consistency of method names etc), but apart from that, this should already represent what we will merge as "step 1" API. I hope to also work a bit on the documentation in parallel with the 0.4 work. (integrate and expand the "overview" above into doc files, as well as include some useful contributions from @reaysawa ).

**Comment by bfredl** [comment]
I wrote some basic documentation for the tree/node API (or rather mostly copied it from api.h .
I plan to merge this quite soon, after that functionality (including test/docs/out-of-the box usability) will be incrementally improved with follow-up PRs. My next priority will be to wrap the new query API https://github.com/tree-sitter/tree-sitter/pull/444 which will form the basis of tree-sitter highlighting (simpler and cleaner than the previous approach #9219 ).

File: ci/build.ps1
Comment (justinmk) [inline ci/build.ps1]: creating things in root (`c:\`) is bad practice and might not work on non-CI machines (depending on user permissions). Although build.ps1 isn't fully usable for non-CI (yet), would be good to avoid making it less so. Could use `build/` or `.deps/` .

File: ci/build.ps1
Comment (justinmk) [inline ci/build.ps1]: this is python 2, might as well use `python3` here since python 2 is EOL and non-CI environments are more likely to have python3 (though ideally we would use powershell instead of python).

File: ci/install.sh
Comment (justinmk) [inline ci/install.sh]: would be nice to put this treesitter setup in a separate function since it is quite involved. (Same for `ci/build.ps1`)

File: ci/run_tests.sh
Comment (justinmk) [inline ci/run_tests.sh]:
```suggestion
export TREESITTER_DIR=$HOME/tree-sitter-build/
```

File: ci/build.ps1
Comment (bfredl) [reply ci/build.ps1]: No, there is exactly nothing about this that "is python2" in any way. This just uses whatever python interpreter that is first in PATH and should work with a wide range of python2 and python3 versions. The point of "python 2 is EOL" that "python" refers to python3, and not that it is somehow forbidden to ever refer to the python interpreter as "python" again. This was a mistake made internally in vim only, and does not apply to shell commands.

File: ci/run_tests.sh
Comment (bfredl) [reply ci/run_tests.sh]: `TREE_SITTER_DIR` is used by the tree-sitter binary, it is not a name chosen by us.

File: ci/build.ps1
Comment (justinmk) [reply ci/build.ps1]: `python` is always python 2 in build.ps1 (see line 83). But I realize now that wouldn't apply for local builds anyway (and powershell would be the ideal approach anyways).

File: runtime/doc/if_lua.txt
Comment (justinmk) [inline runtime/doc/if_lua.txt]:
```suggestion
Currently Nvim does not provide the tree-sitter parsers, instead these must
```

File: ci/install.sh
Comment (bfredl) [reply ci/install.sh]: The rest of the code does not use functions (including python setup and other specific sections which are substantially longer). So I do not know what style or naming convention such a function should be, or what the arguments would be etc. I just wrote code that did the necessary setup here, restructuring the ci setup code seems like a separate PR to me.

File: ci/build.ps1
Comment (bfredl) [reply ci/build.ps1]: I assumed a file in `ci/` is for, well, CI and not a generic build script or something. But I will find some more reasonable path.

File: ci/install.sh
Comment (justinmk) [reply ci/install.sh]: [common/build.sh](https://github.com/neovim/neovim/blob/a7a747ef6948e902929b7bbde4d2140654e73824/ci/common/build.sh) might be a good place.

File: ci/build.ps1
Comment (bfredl) [reply ci/build.ps1]: > python is always python 2 in build.ps1 (see line 83).

Then it is this part that needs to be fixed to prepare for a python3 only future. The code here just assumes that any python is available as "python".

I tried to look for a decompression API in powershell, but I didn't find any that seemed to be "standard" and worked with short code like this. Feel free to suggest/write the right code :)

File: runtime/lua/vim/treesitter.lua
Comment (justinmk) [inline runtime/lua/vim/treesitter.lua]: nit: could use the module pattern instead  e4bd31dbaceefd21932b96f11e7f555d607ffd49 .

File: src/nvim/CMakeLists.txt
Comment (justinmk) [inline src/nvim/CMakeLists.txt]: should we instead delete the amalgamation from the tree? or is it used somewhere?

File: test/functional/lua/treesitter_spec.lua
Comment (justinmk) [inline test/functional/lua/treesitter_spec.lua]:
```suggestion
describe('treesitter API', function()
```

File: test/functional/lua/treesitter_spec.lua
Comment (justinmk) [inline test/functional/lua/treesitter_spec.lua]:
```suggestion
      it("works", function() pending("TREE_SITTER_PATH not set, skipping treesitter parser tests") end)
```

**Review by justinmk — APPROVED** [review]
- I followed the steps in `ci/install.sh` and it worked on my ubuntu env. 🚀 
- ~~Can `src/tree_sitter/` be renamed to `src/treesitter/`?~~

File: ci/build.ps1
Comment (reaysawa) [reply ci/build.ps1]: Some programs rely on environment variables for installation paths without asking for elevated rights, for instance `%APPDATA%`. Maybe [%TMP% or %TEMP%](https://superuser.com/questions/709953/temp-vs-tmp-in-environment-variables) are set at the time the script runs and you could use it.

**Comment by bfredl** [comment]
> Can src/tree_sitter/ be renamed to src/treesitter/?

Not without making changes into multiple tree-sitter source files, something I have tried to avoid (currently the only diff is 5 lines in a single file alloc.h that is not changed by upstream frequently to integrate nvim memory handling).

File: src/nvim/CMakeLists.txt
Comment (bfredl) [reply src/nvim/CMakeLists.txt]: it would make the steps for upgrading tree-sitter sources slightly more complicated, though perhaps it doesn't matter as it is pretty fail-fast if you include the file in an update.

**Comment by maxbrunsfeld** [comment]
If it'd be helpful, I'd be open to a PR that makes `ts_malloc` and friends overridable in some way via the preprocessor, so that you wouldn't have to modify the source code.

**Comment by maxbrunsfeld** [comment]
Actually, it's probably possible already.

I think you could just define `ts_malloc`, `ts_calloc`, `ts_realloc` and `ts_free` as preprocessor macros, and also define the `TREE_SITTER_ALLOC_H_` macro so that the content of `alloc.h` will be skipped. Does that seem reasonable?

**Comment by bfredl** [comment]
@maxbrunsfeld Sure, but it seems to be we can achieve the same result by keeping `alloc.h` as it is now in this PR and just never update it. In general I don't like hiding code in `-Dxxx` compiler flags.

**Comment by maxbrunsfeld** [comment]
Interesting. Just curious - do you generally prefer to directly include dependencies in the repository, vs using a git submodule or some other programmatic way of downloading the dependency repository? I would normally think you'd want to avoid copying the whole tree-sitter source directory into your repository.

**Comment by bfredl** [comment]
@maxbrunsfeld our build system allows for "bundled" dependencies which by default are automatically fetched and built as a preparation step. Though the expectation of these is that these contain a top-level `Makefile` or `CMakeLists.txt` which will build the C library and install it to a specified `$PREFIX`. Distros packaging neovim will typically disable all the bundled builds  and instead link nvim to dynamic libraries packaged separately.

I just vendored the tree-sitter C source for now as it works well enough and seemed the simplest regarding the current packaging situation (and it is not really big, it is smaller than nvim's existing syntax highlighting implementation!). But if tree-sitter were to include a CMakeLists.txt or Makefile or similar to make it easy for distros to build a package for the tree-sitter C library, we could switch to include it as a "bundled" dependency later on.

**Comment by justinmk** [comment]
> But if tree-sitter were to include a CMakeLists.txt or Makefile or similar to make it easy for distros to build a package for the tree-sitter C library, we could switch to include it as a "bundled" dependency later on.

Yes, to be clear: we are waiting on that (is it planned @maxbrunsfeld ?).  But we did not want to be blocked on it meanwhile.

**Comment by maxbrunsfeld** [comment]
> Though the expectation of these is that these contain a top-level `Makefile` or `CMakeLists.txt` which will build the C library and install it to a specified `$PREFIX`

> Yes, to be clear: we are waiting on that (is it planned maxbrunsfeld?)

I hadn't specifically planned on packaging Tree-sitter as a dylib like that, but I'm open to a PR that adds a `Makefile` for that purpose.

**Comment by justinmk** [comment]
What is the  long term recommended way of depending on treesitter, then?

**Comment by maxbrunsfeld** [comment]
I think the simplest approach is to just link it statically into the project, like you are doing currently. In case it's useful, here are links to the build scripts of [`node-tree-sitter`](https://github.com/tree-sitter/node-tree-sitter/blob/355bb66699eb90f6391a9ad6a1027af8bb90f428/binding.gyp#L35-L48) and [`py-tree-sitter`](https://github.com/tree-sitter/py-tree-sitter/blob/7e3eb09b50d0a364076707b4cbb91c3028a81de4/setup.py#L29-L42).

The only things I would usually recommend doing differently are:
* I'd use the amalgamation file `lib.c`, since it saves you from having to determine the list of sources to compile, and compiling `lib.c` is usually faster than compiling the source files separately.
* I would use a git submodule (or just some kind of install script) rather than copying the source files into the parent repository, just because that will make it easier to upgrade Tree-sitter.

That said, I know there are lots of varying opinions and trade-offs related to these issues, so I think what you're doing currently is also totally reasonable.

And like I said, I'm open to adding or modifying Tree-sitter's build setup to make it easier to use. Up until now, because there are so many different ways to build C libraries, I haven't yet had a reason to recommend anything other than ("add `lib.c` to your project, in whatever way you need to").

**Comment by bfredl** [comment]
> I'd use the amalgamation file lib.c, since it saves you from having to determine the list of sources to compile, and compiling lib.c is usually faster than compiling the source files separately.

The amalgation assumes though that utf8proc.c is built as part of the main build, something we don't want. Though this will be solved once/if tree-sitter removes the utf8proc dependency https://github.com/tree-sitter/tree-sitter/pull/436.

> I would use a git submodule (or just some kind of install script) rather than copying the source files into the parent repository, just because that will make it easier to upgrade Tree-sitter.

Sure, I just don't think it is worth it at this point to introduce a third different way of dealing dependencies into our build system (apart from bundled ones and vendored ones), which all downstream packagers need to deal with. Also, a submodule would need to be replaced with something different for our release tarballs anyway. I think we should stick to the vendored approach for now, and before our 0.5 release we could revisit the possibility of turning it to a bundled dependency (probably by us contributing a CMakeLists.txt to tree-sitter), to keep it similar to how we handle other C libraries.

**Comment by blueyed** [comment]
Since we do not plan to ship the vendored source (which is good), I'd rather use a Git submodule until it is bundled already.

While it would be OK during development of this PR it shouldn't really become part of the master branch's Git history.  And with this PR already it would help with not having that code visible here (directly).  Also it makes it easier to carry and upstream patches from there then.

**Comment by justinmk** [comment]
git submodules wiil be a headache, we already decided on the approach in this PR. 

Instead of spending time scripting/supporting submodules, that time could be spent writing a Makfile/CMakeList for TS upstream.

**Comment by milch** [comment]
@bfredl For watchers of this PR, where should we continue following the tree sitter integration work now that this is merged?

**Comment by bfredl** [comment]
I will open the next PR (query API) very soon :)

**Comment by bfredl** [comment]
See #11113. Please continue discussion there or open a new issue for things like build problems.

**Pull request merged**
